src/python/datapreprocessor/datapreprocessor/datanormaliser.py

```diff
@@ -58,3 +68,63 @@ def normalise(file_contents):
     file_contents = clean_minister_column(file_contents)
     file_contents = clean_dates(file_contents)
     return file_contents
+
+def extract_info_from_filename(filename, type_strings=DATA_TYPES):
+    """
+    Look at a filename and attempt to extract the data type and year from it
+
+    Does some simple searching in the filename for certain keywords that
+    indicate what type of data the file may contain. A single data type needs
+    to be found otherwise processing of the file would not be able to continue
+    so an exception is raised in cases where 0 or 2+ types are found.
+    Also looks to see if there's a year string in the filename, this could be
+    either a two or four digit representation. The year is only needed to
+    provide a hint to the date cleaning function later so it's not breaking if
+    it's not there.
+
+    Returns:
+    A dict containing:
+    - type: The data type of the file
+    - year: A year this file may be referring to (or None if ambiguous)
+
+    Raises:
+    - TypeNotFoundException: if no data type identified
+    - MultipleTypesFoundException: if more than one data type identified
+    """
+    type_keys = [type_ for type_ in type_strings if type_ in filename]
+    if not type_keys:
+        raise TypeNotFoundException()
+    elif len(type_keys) > 1:
+        raise MultipleTypesFoundException(type_keys)
+
+    year = None
+    years = []
+
+    # Match four digit runs at the start that begin with 19 or 20
+    matches = re.findall(r'\A((?:19|20)\d{2})(?:\Z|\D)', filename)
+    if matches:
+        years += [int(match) for match in matches]
+
+    # Match four digit runs not at the start (must have a non-digit in front)
+    # that begin with 19 or 20
+    matches = re.findall(r'(?:\D)((?:19|20)\d{2})(?:\Z|\D)', filename)
+    if matches:
+        years += [int(match) for match in matches]
+
+    # Match two digit runs at the start
+    matches = re.findall(r'\A(\d{2})(?:\Z|\D)', filename)
+    if matches:
+        years += [2000 + int(match) for match in matches]
+
+    # Match two digit runs not at the start (must have a non-digit in front)
+    matches = re.findall(r'(?:\D)(\d{2})(?:\Z|\D)', filename)
+    if matches:
+        years += [2000 + int(match) for match in matches]
+
```

Comment:
It is a bit of an edge case but we could also handle two digit years from the 20th century here I think.

For example, the filename 'i_am_quite_an_old_file_98' would produce the year 2098 which is clearly wrong. We could check it against the current year though, and subtract 100 if it is in the future.